numpy/core/tests/test_dtype.py

```diff
@@ -1366,10 +1366,13 @@ def test_metadata(self):
         [type(np.dtype(t)) for t in np.typecodes['All']] +
         [np.dtype(rational), np.dtype])
     def test_pickle_types(self, DType):
-        # Check that DTypes (the classes/types) roundtrip when pickling
+        # Check that DTypes (the classes/types) roundtrip when pickling and
+        # that pickling doesn't change the hash value
+        pre_pickle_hash = hash(DType)
         for proto in range(pickle.HIGHEST_PROTOCOL + 1):
             roundtrip_DType = pickle.loads(pickle.dumps(DType, proto))
             assert roundtrip_DType is DType
+            assert hash(DType) == pre_pickle_hash
```

Comment:
This tests works with the `type(dtype)` not with the `DType`.  The identity check is sufficient here, you need to add a new test that pickles and unpickles the instance (remove the `type()` in the parametrization).

Reply:
Sorry, I didn't understand the point about `type()`.

The bug (hash of `0`) does occur on the dtype itself, not the `rational` type or an instance of `rational`:

```
>>> print(f"{hash(np.dtype(rational))=}")
hash(np.dtype(rational))=0
>>> print(f"{hash(rational)=}")
hash(rational)=8762816130874
>>> print(f"{hash(rational(3, 4))=}")
hash(rational(3, 4))=1966074
```

Reply:
The test is currently set up incorrectly, it's supposed to test whether the type of a dtype is getting pickled correctly. 

Right now, for `rational` and `np.dtype` itself, it's just pickling the instance. 

It should be pickling `type(np.dtype(rational))` and `type(np.dtype)`. Note that the former is not the same thing as `rational`, it's the dtype metaclass for the rational dtype:

```
In [5]: np.dtype(rational)
Out[5]: dtype(rational)

In [6]: type(np.dtype(rational))
Out[6]: numpy.dtype[rational]

In [7]: rational
Out[7]: numpy.core._rational_tests.rational
```

`rational` is the scalar, `dtype(rational)` is the instance, and `numpy.dtype[rational]` is the metaclass. All three are distinct things.

Sebastian is saying that you need to add a new test that tests that pickle round-tripping is correct (e.g. this test, but with the `type()` in the parametrization removed).

Reply:
Hmm, the test doesn't pass anymore with the `type()` removed since the instances don't preserve identity when pickling and unpickling:

```python
>>> dt = numpy.dtype(bool)
>>> pickle.loads(pickle.dumps(dt)) is dt
False

>>> from numpy.core._rational_tests import rational
>>> dt = numpy.dtype(rational)
>>> pickle.loads(pickle.dumps(dt)) is dt
True
```

Is it correct to test equality instead of identity?

---

Also, `type(np.dtype)` is not picklable:

```python
>>> pickle.dumps(type(numpy.dtype))
Traceback (most recent call last):
  ...
PicklingError: Can't pickle <class 'numpy._DTypeMeta'>: attribute lookup _DTypeMeta on numpy failed
```

Is that a separate bug, or expected behavior?

Reply:
Separate issue, maybe it should work, but it's not a big deal because you really never access that metaclass.

Yes, you can test equality, it is a separate issue that it would be nice to return identities more reliably here.  But plase do *not* capitalize `DType`.  The capitalized version looks like the intention is that it is the DType class, but you are testing the dtype instances here.